polling_stations/apps/addressbase/helpers.py

```diff
@@ -232,3 +232,114 @@ def create_address_records_for_council(council, batch_size, logger):
+        self.department_name = department_name
+        self.po_box_number = po_box_number
+        self.sub_building_name = sub_building_name
+        self.building_name = building_name
+        self.building_number = building_number
+        self.dependent_thoroughfare = dependent_thoroughfare
+        self.thoroughfare = thoroughfare
+        self.post_town = post_town
+        self.double_dependent_locality = double_dependent_locality
+        self.dependent_locality = dependent_locality
+        self.address_label = []
+
+    def generate_address_label(self):
+        """Construct a list for address label.
+
+        Non-empty premises elements are appended to the address label in the
+        order of organisation_name, department_name, po_box_number (which
+        must be prepended with 'PO Box', sub_building_name, building_name,
+        building_number, then the rest of the elements except for Town and
+        Postcode because we want them in their own fields. This isn't strict
+        address label but we're probably loading them into a database.
+        """
+        if self.organisation_name:
+            self.address_label.append(self.organisation_name)
+        if self.department_name:
+            self.address_label.append(self.department_name)
+        if self.po_box_number:
+            self.address_label.append('PO Box ' + self.po_box_number)
+
+        elements = [
+                self.sub_building_name,
+                self.building_name,
+                self.building_number,
+                self.dependent_thoroughfare,
+                self.thoroughfare,
+                self.double_dependent_locality,
+                self.dependent_locality,
+        ]
+
+        for element in elements:
+            if element:
+                self._append_to_label(element)
+
+        # pad label to length of 7 if not already
+        if len(self.address_label) < 7:
+            for i in range(7 - len(self.address_label)):
+                self.address_label.append('')
+
+        # finally, add post town
+        self.address_label[5] = self.post_town
+
+        return ", ".join([f for f in self.address_label if f])
+
+    def _is_exception_rule(self, element):
+        """ Check for "exception rule".
+
+        Address elements will be appended onto a new line on the lable except
+        for when the penultimate lable line fulfils certain criteria, in which
+        case the element will be concatenated onto the penultimate line. This
+        method checks for those criteria.
+
+        i) First and last characters of the Building Name are numeric
+          (eg '1to1' or '100:1')
+        ii) First and penultimate characters are numeric, last character is
+          alphabetic (eg '12A')
+        iii) Building Name has only one character (eg 'A')
+        """
+        if element[0].isdigit() and element[-1].isdigit():
+            return True
+        if element[0].isdigit() and element[-2].isdigit() and element[-1].isalpha():
```

Comment:
Might `element[-2]` raise an `IndexError`?

Reply:
Good question/point - I probably need to add a `len()` check here.